---
title: "genophenoR: An R package for querying and analyzing EHR data through the BD2K PIC-SURE RESTful API"
author:
- name: Alba Gutierrez-Sacristan
  affiliation: 
  - &dbmi Department of Biomedical Informatics, Harvard Medical School 
  email: alba_gutierrez@hms.harvard.edu
- name: Romain Guedj
  affiliation: *dbmi
- name: Laura I. Furlong
  affiliation: 
    - &grib Research Programme on Biomedical Informatics (GRIB)
- name: Paul Avillach
  affiliation: *dbmi
date: "`r doc_date()`"
package: "`r pkg_ver('genophenoR')`"
bibliography: general_overview.bib
csl: biomed-central.csl
vignette: >
  %\VignetteIndexEntry{genophenoR: An R package for querying and analyzing EHR data through the BD2K PIC-SURE RESTful API}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document2:
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

The `genophenoR` package contains functions to query different databases through the BD2K RESTful API, a programmatic interface that provides access to different data sources, making easier data accessibility, analysis reproducibility and scalability. The `genophenoR` package includes analysis functions to study demographic population, phenoty descriptive analysi, prevalence and co-occurrence, according to patient age, gender and mutation status if data available. A special focus is made on visualization of the results, providing a variety of representation formats such as networks, heatmaps and barplots (Table \@ref(tab:viz-opt)).


## Background

In the era of big data and precision medicine, the number of databases containing clinical and genomic information is increasing in an exponantial way. Enables the experts to focus on their research questions rather than in the computation data mergin, access and management is one of the biggest challenges nowadays. 

Towards the aim of developing reproducible and scalable analysis, the `genophenoR` package has been developed. This R package allows to make queries to different databases through the PIC-SURE API, simplifying the data connection and analysis. Moreover it allows to easily reproducible analysis between different databases as well as its integra-tion with other packages available in R to develop bioinformatics analy-sis workflows. 

The tasks that can be performed with `genophenoR` package are the following:

  1. query through the BD2K PIC-SURE API RESTful API
  2. Demographic descriptive analysis based on age and gender
  3. Phenotype descriptive analysis taking into account continuous and categorical variables. 
  4. Phenotype prevalence and co-occurrence analysis according to age, gender and gene mutation status if available. 
  5. Extract those patients that have been diagnosed with a determined pair of phenotypes. 

  
In the following sections the specific functions that can be used to address each of these tasks are presented.

## Installation 

The package \texttt{genophenoR} is provided through GitHub. To install \texttt{genophenoR} the user must type the two following commands in an R session:

```{r bioC, eval=FALSE}
install_github("hms-dbmi/genophenoR")
```

```{r knirt, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
```

```{r load_library, message=FALSE, warning=FALSE}
library(genophenoR)
```


## S4 objects 

### `genopheno`

`genopheno` object is obtained when `queryGenoPheno` and `genophenoZscore` functions are applied. This object is used as input for `demographicSummary` and `phenotypeSummary`  descriptive functions. The `genopheno` object contains the table dataset and shows the main characteristic of our dataset:

  - The number of mutations variables (N. Mutations)
  - The number of phenotypes variables (N. Phenotypes)
  - The number of patients (N. Patients)

```{r genophenoObj1, echo=FALSE}
load(system.file("extdata", "genophenoExData.RData", package="genophenoR"))
```

```{r genophenoObj2}
genophenoExData
class( genophenoExData )
```

This object comes with a series of functions to allow users to interact with the information retrieved. These functions are `nphenotype`, `nmutation` and `extract`. The first function `nphenotype` returns the phenotypes present in the output of the given query. `nmutation` is the homologous function but for the mutations. Finally, the function `extract` returns a formatted `data.frame` with the complete set of information contained in the object, the query results. 

### `genophenoComor`

`genophenoComor` object is obtained when `genoPhenoComorbidity` function is applied. This object is used as input for `genoPhenoPrevalence`, `genoPhenoHeatmap`and `genoPhenoNetwork`  functions. The `genophenoComor` object contains the comorbidity results and shows the main specifications of the comorbidity analysis:

  - Minimum age of the patients in the dataset (Age Min)
  - Minimum age of the patients in the dataset (Age Max)
  - Patients gender
  - Mutation, if selected
  
And a summary of the results:

  - Number of patients in the age and gender interval in the database\
  - Number of patients in the age and gender interval with the mutation select
  - The prevalence of the mutation in the dataset
  - Odds ratio range 
  - Relative risk range
  - Phi value range
  - Number of resultant comorbidities

```{r genophenoComorObj1, echo=FALSE}
load(system.file("extdata", "genophenoComor.RData", package="genophenoR"))
```

```{r genophenoComorObj2}
genophenoComor
class( genophenoComor )
```

The `extract` function can be applied to `genophenoComor` objects. It returns a formatted `data.frame` with the comorbidity results contained in the object. 

# Retrieve data from the database

`genophenoR` package allows exploring the database information using the specifc url, the key and a JSON query in text format. It retrieves the information that is available in the database according to the user query and allows the analysis and results visualization in different  ways.

In order to perform a query in a database, the user can apply the `irctquerty` function. This function retrieves database information using:

  - url
  - key
  - query


As an example data from NHANES database [@patel2016database] will be retrieve applying the `irctquerty` function to:

  - __url__: "https://nhanes.hms.harvard.edu/"
  - __key__:to access to the data
  - __JSON query__: _ select the variables pcb153, HIV, herpes, patient age and gender from all those patients in NHANES database that have age information_. The example of the JSON query provided as input is shown in Figure 1. 


```{r irctquery, eval= FALSE, message=FALSE, warning=FALSE}
query <- irctquery( 
              url         = "https://nhanes.hms.harvard.edu/", 
              apiKey      = "a77l42anvcgdtkfcvbl7hnp2v9", 
              query       = system.file("extdata", "jsonQueryNhanes", package="genophenoR"))
              )
```

Figure 1, Query example in NHANES database: 
![wssss](C:/Users/AG440/Desktop/genophenoR/inst/queryNhanes.png)

The result is a `data.frame` object:

```{r irctExample, eval = FALSE}
class( query )
kable( head( query ) )
```

In the particular example used, by inspecting the `data.frame` object, we can see that there are 41474 patients in the database that have age data, and the pcb153, hiv and herpes information for these patients have been extracted correctly. 

    

# Transform the data in a genopheno object 



```{r convertToGenophenoRobject}
genophenoExData <- queryGenoPheno( inputDataFile = paste0(system.file("extdata", package="genophenoR"), "/queryOutput.txt"), 
                                   verbose       = TRUE )
genophenoExData
head( genophenoR::extract( genophenoExData ) )
```

# Transform continuous in categorical variables: zscore


```{r zscore}
zscoreExample <- genophenoZscore( input      = genophenoExData,
                                  cutOff     = c( -2, 2 ), 
                                  nfactor    = 10, 
                                  verbose    = TRUE )
zscoreExample
head( genophenoR::extract( zscoreExample ) )
```


# Demographic analysis 

The function `demographicSummary` describes the demographic characteristics (sex, age) of the population under study.
The `demographicSummary` function requires three arguments:

  - input: the genopheno object.
  - maleCode: the symbol which denotes males in the table (e.g., 0, M, Male...etc)
  - femaleCode: the symbol which denotes females in the table (e.g., 1, F, Female...etc)


The `demographicSummary` function output includes:
  
  - A barplot with the age distribution of the whole study population
  - A boxplot showing the age distribution by gender.
  - A pie chart representing the gender distribution.

```{r demographicAnalysis, fig.width=8, fig.height=8}
genophenoR::demographicSummary ( input       = genophenoExData, 
                                 maleCode    = "male", 
                                 femaleCode  ="female", 
                                 verbose     = TRUE )
```


# Phenotype analysis

The function `phenotypeSummary` describes the phenotypic characteristics for the whole study population then according to the status regarding one selected gene.    

The `phenotypeSummary` function requires 2 arguments: 

  - input: the `genopheno` object.
  - mutation: the mutation of interest
  - ifile: to create a file summarizing the information displayed in the barplots. It will be saved in your working directory.
  - showTable: the table is displayed before the barplot. 

The `phenotypeSummary` function output includes:   
  
  - A barplot for each categorical phenotype variable
  - A boxplot for each continuous phenotype variable
  - A table if the showTable argument is TRUE

```{r phenosumm, fig.width=8, fig.height=8, warning= FALSE}
phenotypeSummary( input       = genophenoExData,
                  showTable   = TRUE, 
                  showFigures = FALSE)
```


# Phenotypes co-occurrence and prevalence using 

The `genoPhenoComorbidity` identifies all the possible phenotype co-ocurrence and quantify them.  Five different quantification measures are estimated:   

  - fisher test with Benjamini-Hochberg FDR [@hidalgo2009dynamic]
  - comorbidity score [@roque2011using] 
  - relative risk [@hidalgo2009dynamic]
  - Pearson correlation   
  - odds ratio. 

According to the type of study, the user can select the co-ocurence measures needed as well as the cut-off value. As a result of the phenotype co-occurrence analysis, a table containing the different estimated measurements is obtained. 


The `genoPhenoComorbidity` function requires 6 arguments:  

  - input: the `genopheno` object
  - pth: the path where the file with the phenotype values generated previously is located.
  - aggregate: if all the possible phenotypes values want to be considered individually, aggregate must be FALSE. Otherwhise, the phenotype values should be manually completed by the user (column yes/no) and aggregate must be TRUE. 
  - ageRange: determines what is the age range of interest for performing the analysis.
  - gender: determine what is the gender of interest for performing the co-occurrence analysis. 
  - mutation: determine the mutation of interest and the value of this mutation.


## Phenotype co-occurrence

```{r phenoCooccurrence}
genophenoComor <- genoPhenoComorbidity( 
              input         = genophenoExData,
              pth           = system.file("extdata", package="genophenoR"), 
              aggregate    = FALSE, 
              ageRange      = c(4,18),
              gender        = "ALL", 
              )
genophenoComor
```

## Prevalence

The function `genophenoPrevalence`  displays in a data frame:
    * Phenotype
    * Number of patients (N_patients)
    * Prevalence
    * Confidence interval



```{r genophenoPrevalence}
prevalence <- genoPhenoPrevalence( input = genophenoComor )

kable( prevalence, row.names = FALSE)
```

## Ploting the co-occurrence results

`genophenoR` offers several options to visualize the results from the phenotype co-occurrence analysis, in networks and heatmaps.

### Network
By applying the `genophenoNetwork` function. A network showing the phenotypes comorbidities is obtained. By default `genophenoR` package shows a network with `layout.fruchterman.reingold` layout where the phenotypes are joined according to the number of patients that have both phenotypes ( selectValue = patientsPhenoAB). The user can change this value to the different comorbidity measures - fisher, oddsRatio, relativeRisk, phi, expect, score, fdr, PercentagePhenoAB- that are estimated for each pair of phenotypes. 

The `genoPhenoNetwork` function requires 3 arguments as input:  

   - input: the `genophenoComor` object
   - selectValue: the co-occurence measurement 
   - cutOff: the numeral value of the cutOff

The `genoPhenoNetowrk` function output is a network. Node size is proportional to the phenotype prevalence. Edge numbers represent the selected value. 


```{r phenoNetwork, fig.width=10, fig.height=10}
genoPhenoNetwork ( input       = genophenoComor,
                   selectValue = "PercentagePhenoAB", 
                   cutOff      = 10)
```


### Heatmap
By applying the `genophenoHeatmap` function, a heatmap is displayed. This function is homologous to the `genophenoNetwork` one.  The heatmap represents the selected value that by default will be `patientsPhenoAB`. The user can change this value to the different comorbidity measures - fisher, oddsRatio, relativeRisk, phi, expect, score, fdr, PercentagePhenoAB- that are estimated for each pair of phenotypes. 

The `genoPhenoHeatmap` function requires 3 arguments:  

  - input: the `genophenoComor` object
  - selectValue: the co-occurence measurement 
  - cutOff: the numeral value of the cutOff

The `genoPhenoHeatmap` function output is a heatmap. Blue color represents the lower values,  red color represets the upper values. The co-occurence measurement value is showed in each cell. 

```{r phenoHeatmap, fig.width=10, fig.height=10 }
genoPhenoHeatmap ( input       = genophenoComor,
                   selectValue = "PercentagePhenoAB", 
                   cutOff      = 0)
```

## Ploting the prevalence results

```{r phenoPrevalenceNetwork, fig.width=10, fig.height=10 }
prevalenceNetwork( input   = genophenoComor, 
                   layout  = "layout.fruchterman.reingold", 
                   title   = "Phenotype prevalence network", 
                   verbose = TRUE
              )
```


# Summary of `genophenoR` functions available

| Input Object    | `genophenoR` function | Output Generated                                  |
| ----------------|-----------------------|---------------------------------------------------|
| `genopheno`     | `demographicSummary`  | 3 plots: age and gender distribution, relation between age and gender distribution|
|                 | `genoPhenoComorbidity`| It generates a genophenoComor object              |
|                 | `genoPhenoPatientsSelection`| It retrieves a list of patients suffering two phenotypes of interest |
|                 | `genoPhenoComorbidity`| It generates a genophenoComor object              |
|                 | `extract`             | It generates a data.frame containing the raw data from a query |
|                 | `nmutation`           | It retrieves the mutations of your data |
|                 | `nphenotype`           | It retrieves the phenotypes of your data |
| genophenoComor  | `genoPhenoHeatmap`    | It generates a heatmap showing the comorbidity analysis results |
|                 | `genoPhenoNetwork`    | It generates a netowork showing the comorbidity analysis results |
|                 | `extract`             | It generates a data.frame containing the comorbidity results |

: (\#tab:viz-opt) Functions in `genophenoR` R package



# Bibliography
